config/machine_id.go

```diff
@@ -0,0 +1,237 @@
+package config
+
+import (
+	"crypto/rand"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	"github.com/google/uuid"
+)
+
+type MachineIDGetter interface {
+	GetMachineID() uuid.UUID
+}
+
+type FileReader func(fileName string) ([]byte, error)
+type HostNameReader func() (name string, err error)
+
+type generatorFn func() ([]byte, error)
+
+// Used to create a bitmask to identify how the machine ID was generated
+const (
+	isSnapEnv     = 1
+	etcMachineID  = 1 << 1
+	dbusMachineID = 1 << 2
+	dmiProductID  = 1 << 3
+	cpuSerial     = 1 << 4
+	boardSerial   = 1 << 5
+	productUUID   = 1 << 6
+)
+
+type MachineID struct {
+	hostNameReader HostNameReader
+	fileReader     FileReader
+
+	sync.Mutex
+	machineID    *uuid.UUID // store computed device ID
+	usedInfoMask int16      // stores the bitmask to identify how the ID was generate
+}
+
+func NewMachineID(fileReader FileReader, hostNameReader HostNameReader) *MachineID {
+	return &MachineID{
+		hostNameReader: hostNameReader,
+		fileReader:     fileReader,
+		machineID:      nil,
+		usedInfoMask:   -1,
+	}
+}
+
+func (getter *MachineID) GetMachineID() (ret uuid.UUID) {
+	getter.Lock()
+	defer getter.Unlock()
+
+	return getter.getMachineID()
+}
+
+// Return the bitmask to identify what information was used to generate the machine ID
+func (getter *MachineID) GetUsedInformationMask() int16 {
+	getter.Lock()
+	defer getter.Unlock()
+
+	if getter.machineID != nil {
+		getter.getMachineID()
+	}
+
+	return getter.usedInfoMask
+}
+
+func (getter *MachineID) getMachineID() (ret uuid.UUID) {
+	if getter.machineID != nil {
+		return *getter.machineID
+	}
+
+	defer func() {
+		if getter.machineID == nil {
+			getter.machineID = &ret
+			if IsUnderSnap() {
+				getter.usedInfoMask |= isSnapEnv
+			}
+		}
+	}()
+
+	getter.usedInfoMask = 0
+
+	id, err := getter.generateID()
```

Comment:
[small] Do we want to log the errors here? It may be tricky to figure out that something is wrong if we'll always return the fallback id and we could miss that the original methods are failing